### PR TITLE
Feature/ Better LoRA : Dropout, Conv2d

### PR DESCRIPTION
**Pull request by cloneofsimo**
So unlike classical LLMs, LoRA also has other many modules as well. Arguably, it seems like many "important" features come from Resnet features. This is clearly demonstrated by, for example, plug-and-play prior

image
> Check out the details on this paper : https://pnp-diffusion.github.io/

Natural question to ask is : Does dreambooth yield fine-grained level of details, such as eyes or skins, because it is able to tune resnets? Is Q, K, V, O simply not enough?

In this PR I will try to answer this questions with bunch of experiments.

Here are some initial results : 

This result is LoRA rank 4, with "CrossAttention", "Attention", "GEGLU" Replaced

pr_no

Now, this is LoRA rank 1, with "ResnetBlock2D", "CrossAttention", "Attention", "GEGLU" Replaced. Now, number of parameters are now about as 2 times higher : as high as 9.7MB. Of which, only 2MB is LoRA of  "CrossAttention", "Attention", "GEGLU", so I suspect that if it helps, we might make LoRA of  TR rank 4 and LoRA of Resnet rank 1.

pr_res

All trained with same number of steps, sampled with same parameters. This case, it looks like its a tie. I'll try on other models as well. I think it is a good time to start implementing fidelity metrics as well, instead of CLIP alignment score.

### CODE REVIEW

**Comment by brian6091** [comment]
Excellent!

**Comment by cloneofsimo** [comment]
Been testing with rank 4 of resnet as well : but the result seems only marginally better. Of course this is totally subjective so it just doesn't worth the size increase.
Next step would be compare all sets : MLP, Attention, Resnet, train them one at a time to compare the results.

image

**Comment by hafriedlander** [comment]
Definitely interesting experiments.

What parameter are we optimizing for, total binary size, training speed, multiple concept accuracy, something else?

(The reason I ask is my first thought was "how does this compare to just rank 16 on CrossAttention / Attention / GEGLU")

Is there some way we can auto-detect when the rank is insufficient? (Like maybe a flat gradient when training while still having a high error rate).

**Comment by hafriedlander** [comment]
Also, just from a end-user POV, the greatest strength of LoRA for me is the easy adjust-ability of strength of the various adjustments. It'd be interesting to see how useful post-training adjustment of individual model LoRA weights was.

**Comment by cloneofsimo** [comment]
So I think it is kindof constraint optimization at this point. We just don't want output too large, but quality as dreambooth as high as possible. But the objective is also mixed. distortion, perceptual fidelity, and editability is all the performance we desire, but they have tradeoff relationship. perceptual fidelity is also kinda ill-defined, as CLIP score doesn't seem to represent that very well, unlike custom diffusion or textual inversion paper would like to suggest.

**Comment by cloneofsimo** [comment]
It seems like what many people are looking for can be simply described : identity preservation + editability.
Community seems to have high preference for usage on faces, and thus facial identity preservation is one thing they really want.
Second is that it has to have high editability as well, which I think CLIP score can well describe.

**Comment by cloneofsimo** [comment]
Now @brian6091 and I had an idea on making the LoRA on the resent part only on the upsample unet layers, because downsample parts are kinda used to compress representation and not to generate with fidelity. So we'll see if it works better. @brian6091 are you going to leave a PR or it just on your own thing?

**Comment by cloneofsimo** [comment]
I've added dropout as well so this PR is no longer only about conv layer. I'll rename it

**Comment by cloneofsimo** [comment]
Looks like adding dropout helps! Just like in the paper, I've used dropout with 0.1. Same seed, everything.

pr_res_dp

**Comment by ExponentialML** [comment]
Your work and contributions are very underrated. Great stuff @cloneofsimo!

**Comment by brian6091** [comment]
Yeah, I think the ability to do ablation experiments will be super interesting. There are subjective differences between the different components (CrossAttention, FFN for example) that may be hard to capture with objective metrics. But from the end-user perspective I think being able to define your own objective and having the tools to achieve that is ideal.

**Comment by brian6091** [comment]
I'll work the scale/nonlinearity code back in once you've stabilized this. Also subtle effects, but worth it IMO for the trivial cost.

**Comment by brian6091** [comment]
> Now @brian6091 and I had an idea on making the LoRA on the resent part only on the upsample unet layers, because downsample parts are kinda used to compress representation and not to generate with fidelity. So we'll see if it works better. @brian6091 are you going to leave a PR or it just on your own thing?

I can leave a PR.

**Comment by cloneofsimo** [comment]
Done various experiments, but more needs to be done. I've got mixed results in my case, so i'll add these options as optional for now. using dropout does make a difference though.

**Comment by cloneofsimo** [comment]
# Note:

I've found that having resnet trained requires very low learning rate : something like 5e-6 for me.

**Comment by okaris** [comment]
@cloneofsimo First of all thank you for creating this repo. I've been tinkering with LoRA but can't say it's faster than (let alone twice as fast) than Dreambooth. Could you share the args you used for the above examples please?

**Comment by cloneofsimo** [comment]
Is this related to conv loras? Or just lora pti in general? @okaris

**Comment by okaris** [comment]
The comment, lora pti in general. I can open a new issue for that. The question, I'd really like to know the settings you used for the above trainings and how long they took. Thanks!

**Comment by cloneofsimo** [comment]
It's been a while since I've evaluated time to train these models, but these take  < 6min in general. I think they aren't as fast as the previous ones (currently in training scripts folder), because lora_pti is not optimized for speed and memory since no 8bit adam + xformers are tested. It is the textual inversion part that takes a long time since they are currently done with full precision.

**Comment by cloneofsimo** [comment]
I am continuing to optimize for perceptual performance first, and README is bit misguiding because they were not based on lora_pti scripts. Better fix that.

**Comment by ExponentialML** [comment]
Hey @cloneofsimo. Is `nn.Conv2d` needed here?
        https://github.com/cloneofsimo/lora/blob/583b1e73421a061975f4557aae3f786d289fdd23/lora_diffusion/lora.py#L565

**Comment by cloneofsimo** [comment]
Ah, these + also other tools are currently unsupported for lora conv2d. Rests are coming in as a feature soon.
Meanwhile, patch_pipe will take care with no problem

**Comment by ExponentialML** [comment]
Sweet, thanks!
